contracts/protocol/facets/SequentialCommitHandlerFacet.sol

```diff
@@ -0,0 +1,195 @@
+        }
+
+        if (_priceDiscovery.side == Side.Bid) {
+            require(seller == msgSender(), NOT_VOUCHER_HOLDER);
+        }
+
+        // First call price discovery and get actual price
+        // It might be lower tha submitted for buy orders and higher for sell orders
+        uint256 actualPrice = fulFilOrder(_exchangeId, tokenAddress, _priceDiscovery, _buyer, offer.sellerId);
+
+        // Calculate the amount to be kept in escrow
+        uint256 escrowAmount;
+        {
+            // Get sequential commits for this exchange
+            SequentialCommit[] storage sequentialCommits = protocolEntities().sequentialCommits[_exchangeId];
+
+            {
+                // Calculate fees
+                uint256 protocolFeeAmount = tokenAddress == protocolAddresses().token
+                    ? protocolFees().flatBoson
+                    : (protocolFees().percentage * actualPrice) / 10000;
+
+                // Calculate royalties
+                (, uint256 royaltyAmount) = IBosonVoucher(protocolLookups().cloneAddress[offer.sellerId]).royaltyInfo(
+                    _exchangeId,
+                    actualPrice
+                );
+
+                // Verify that fees and royalties are not higher than the price.
+                require((protocolFeeAmount + royaltyAmount) <= actualPrice, FEE_AMOUNT_TOO_HIGH);
+
+                // Get price paid by current buyer
+                uint256 len = sequentialCommits.length;
+                uint256 currentPrice = len == 0 ? offer.price : sequentialCommits[len - 1].price;
+
+                // Calculate the minimal amount to be kept in the escrow
+                escrowAmount = Math.max(actualPrice, protocolFeeAmount + royaltyAmount + currentPrice) - currentPrice;
+
+                // Update sequential commit
+                sequentialCommits.push(
+                    SequentialCommit({
+                        resellerId: buyerId,
+                        price: actualPrice,
+                        protocolFeeAmount: protocolFeeAmount,
+                        royaltyAmount: royaltyAmount
+                    })
+                );
+            }
+
+            // Make sure enough get escrowed
+            if (_priceDiscovery.side == Side.Ask) {
+                if (escrowAmount > 0) {
+                    // Price discovery should send funds to the seller
+                    // Nothing in escrow, need to pull everything from seller
+                    if (tokenAddress == address(0)) {
+                        // If exchange is native currency, seller cannot directly approve protocol to transfer funds
+                        // They need to approve wrapper contract, so protocol can pull funds from wrapper
+                        // But since protocol otherwise normally operates with native currency, needs to unwrap it (i.e. withdraw)
+                        FundsLib.transferFundsToProtocol(address(weth), seller, escrowAmount);
```

Comment:
"Seller in the middle" is not a viable option, since it would require that seller approves voucher transfer to the protocol. But that would allow the malicious buyer to create a malicious priceDiscovery contract which could steal the voucher.